src/main/flight/failsafe.c

```diff
@@ -263,22 +272,24 @@ void failsafeUpdateState(void)
                             // Drop the craft
                             failsafeActivate();
                             failsafeState.phase = FAILSAFE_LANDED;      // skip auto-landing procedure
-                            failsafeState.receivingRxDataPeriodPreset = PERIOD_OF_3_SECONDS; // require 3 seconds of valid rxData
                             break;
+#ifdef USE_GPS_RESCUE
                         case FAILSAFE_PROCEDURE_GPS_RESCUE:
                             failsafeActivate();
                             failsafeState.phase = FAILSAFE_GPS_RESCUE;
-                            failsafeState.receivingRxDataPeriodPreset = PERIOD_OF_3_SECONDS;
                             break;
+#endif
                     }
                 }
                 reprocessState = true;
                 break;
 
             case FAILSAFE_LANDING:
                 if (receivingRxData) {
-                    failsafeState.phase = FAILSAFE_RX_LOSS_RECOVERED;
-                    reprocessState = true;
+                    if ((failsafeState.validRxDataReceivedAt - failsafeState.validRxDataFailedAt) > (failsafeConfig()->failsafe_recovery_delay * MILLIS_PER_TENTH_SECOND + PERIOD_RXDATA_RECOVERY)) {
+                        failsafeState.phase = FAILSAFE_RX_LOSS_RECOVERED;
+                        reprocessState = true;
+                    }
```

Comment:
I'm not sure we want to enforce the recovery delay here. Why would we want to prevent the pilot from immediately regaining control for the `AUTO-LAND` procedure?

And since they shouldn't apply to anything but GPS Rescue failsafe behavior, the parameters should be renamed to be related to `gps_rescue` and moved in the GPS Rescue PG rather than be part of `failsafeConfig`.

Reply:
I think the same logic why recovery logic is added to Rescue also applies to auto land. Why not? Now that you've mentioned, I've checked the inav code and it indeed applies to all recovery procedures.